### ISSUE TITLE
Footnote labels may contain spaces but not newlines

### ISSUE THREAD

**Issue by hellux** [jgm/djot.js#146]
Reference labels for links or footnotes may contain spaces:

```
[text][link lbl]
[^footnote lbl]

[link lbl]: url
```
```html
<p><a href="url">text</a>
<a id="fnref1" href="#fn1" role="doc-noteref"><sup>1</sup></a></p>
<section role="doc-endnotes">
<hr>
<ol>
<li id="fn1">
<p><a href="#fnref1" role="doc-backlink">↩︎︎</a></p>
</li>
</ol>
</section>
```

And labels for links may contain newlines but labels for footnotes may not:
```
[text][link
lbl]
[^footnote
lbl]

[link lbl]: url
```
```html
<p><a href="url">text</a>
[^footnote
lbl]</p>
```
This seems to contradict goal 7, and perhaps goal 8. If e.g. a text editor wraps the line in the middle of a footnote label it will no longer be parsed as a footnote.

Edit: meant to post this in jgm/djot.js (can issues be moved?). However, djot.lua seems to behave in the same way.

**Comment by dereuromark** [comment]
Good point.
Had to fix that in PHP too: https://github.com/php-collective/djot-php/pull/267

I think the current single-line rule for footnote labels is the better behavior, even though link-reference labels may span lines.

I would suggest documenting this constraint explicitly in djot rather than broadening footnote labels to multiple lines. If multiline footnote labels were desired, the definition syntax and label-normalization rules would first need to be extended together; changing only inline references would leave references that cannot be defined.

**Comment by jgm** [comment]
I think that note labels and link labels should work the same; if note labels allow spaces, they should allow line breaks too.
Otherwise it's an odd difference in behavior you have to keep in your head. In addition, automatic wrapping tools may wrap on spaces inside note labels.

**Comment by jgm** [comment]
Looks like newlines are allowed:
```
% ./lib/cli.js --to html  
[^a
b]

[^a
b]:  
    foo
<p><a id="fnref1" href="#fn1" role="doc-noteref"><sup>1</sup></a></p>
<section role="doc-endnotes">
<hr>
<ol>
<li id="fn1">
<p>b]:  
foo<a href="#fnref1" role="doc-backlink">↩︎</a></p>
</li>
</ol>
</section>
```
```
% ./lib/cli.js --to html
[^a
b]

[^a b]: 
    foo
<p><a id="fnref1" href="#fn1" role="doc-noteref"><sup>1</sup></a></p>
<section role="doc-endnotes">
<hr>
<ol>
<li id="fn1">
<p>foo<a href="#fnref1" role="doc-backlink">↩︎</a></p>
</li>
</ol>
</section>
```
```
% ./lib/cli.js --to html
[text][link
lbl]
[^footnote
lbl]

[link lbl]: url
<p><a href="url">text</a>
<a id="fnref1" href="#fn1" role="doc-noteref"><sup>1</sup></a></p>
<section role="doc-endnotes">
<hr>
<ol>
<li id="fn1">
<p><a href="#fnref1" role="doc-backlink">↩︎</a></p>
</li>
</ol>
</section>
```
So, can this be closed?

**Comment by dereuromark** [comment]
I do not think the first two examples show that newlines are allowed - I think
they show a bug in the definition scanner, and the HTML output hides it.

Two things are worth separating.

**Reference side: newlines really are accepted now, and normalize to a space.**

Input:

```
[^a
b]

[^a b]: foo
```

The reference's label comes out as `a b` and binds to the one-line definition
(`"footnote_reference": { "text": "a b" }` in the AST). That is a behavior
change since this issue was filed - the original report has `[^footnote` /
`lbl]` staying literal - and it does address the goal 7 complaint on the
reference side: an editor wrapping a reference no longer breaks it.

**Definition side: the newline is not supported, it leaks.**

`pattFootnoteStart` is

```js
const pattFootnoteStart = pattern("\\[\\^([^\\]]+)\\]:[ \\t\\r\\n]");
```

`[^\]]+` admits a newline, and `find` runs against the whole subject rather
than the current line, so the label capture crosses the line break. But the
block loop is still line-oriented, so the remaining physical lines of the
definition are re-parsed as note content. Your own first example shows it: the
note body is

```
b]:  
foo
```

i.e. the second line of the definition is duplicated into the note's own
content. Reduced, with the reference written on one line so only the
definition is in question:

Input:

```
[^a b]

[^a
b]:  
    foo
```

Output:

```html
<p><a id="fnref1" href="#fn1" role="doc-noteref"><sup>1</sup></a></p>
<section role="doc-endnotes">
<hr>
<ol>
<li id="fn1">
<p>b]:  
foo<a href="#fnref1" role="doc-backlink">↩︎</a></p>
</li>
</ol>
</section>
```

One reason this is easy to miss when reading the HTML: an unresolved footnote
reference still renders a numbered note with a backlink, so "a note appeared"
is not evidence that the reference resolved. With no definition at all:

Input:

```
[^a
b]

tail
```

Output:

```html
<p><a id="fnref1" href="#fn1" role="doc-noteref"><sup>1</sup></a></p>
<p>tail</p>
<section role="doc-endnotes">
<hr>
<ol>
<li id="fn1">
<p><a href="#fnref1" role="doc-backlink">↩︎</a></p>
</li>
</ol>
</section>
```

So I would not close on the current behavior. There seem to be two coherent end
states:

1. Labels are physical-line identifiers on both sides: tighten the capture to
   `[^\]\r\n]+`, and specify that a reference may still wrap because the soft
   break normalizes to a space. This keeps the goal 7 property for references
   and removes the leak.
2. Multiline definition labels are genuinely supported, in which case the
   scanner has to consume through the marker's real end position instead of
   leaving those lines to the block loop, and the label needs a defined
   normalization for matching.

Either way the definition side needs a change, so I do not think the current
state closes it.

One further data point, since djot.lua was mentioned. djot-php (which I
maintain) landed the strict reading earlier today - a label is a physical-line
identifier on both sides, so `[^a` / `b]` stays literal - which means it now
diverges from djot.js on the reference side, in the direction of option 1
minus the normalization. I am happy to move it either way, but the part worth
pinning explicitly is whether a reference label is normalized before lookup:
that is the difference between a wrapped reference binding to `[^a b]:` and
not binding at all, and it is currently implied by djot.js's behavior rather
than stated anywhere.

**Comment by jgm** [comment]
OK, right, I was only looking at the inline parser (for references); I didn't look at the block parser (which is where you find `pattFootnoteStart`).

I think that the behavior we want (to be consistent with the way reference links are treated) is: allow line breaks in reference links and footnote references (inline parser), but disallow line breaks in reference labels and footnote labels (block parser).

The inline parser already handles line breaks inside footnote references, as demonstrated.  So I think we just need to tighted up `pattFootnoteStart` in the block parser. (This is your solution 1, I think.)

**Comment by dereuromark** [comment]
Yes, that split is the right one. The definition side is where the leak is, and tightening `pattFootnoteStart` to exclude `\r\n` removes it without touching the reference side.

One thing that is worth writing down as part of this, because the fix is what makes
it load-bearing: **label normalization before lookup.**

Once the block parser can no longer produce a multiline label, a wrapped reference
can only resolve if the label is normalized. That is what makes
`[^a` / `b]` find `[^a b]:` rather than dangle. Today that behavior exists but the
rule is not stated in the syntax description, so an implementation reading the spec
alone has no way to arrive at it.

It is not a new rule. Measured on the released 0.3.2, normalization is already there
and already broader than the newline case:

| input | reference | definition | binds |
| --- | --- | --- | --- |
| double space | `[^a  b]` | `[^a b]:` | yes |
| tab | `[^a<TAB>b]` | `[^a b]:` | yes |
| leading/trailing space | `[^ a ]` | `[^a]:` | yes |
| double space on the definition side | `[^a b]` | `[^a  b]:` | yes |
| newline | `[^a` / `b]` | `[^a b]:` | no, in 0.3.2 |

Link references behave the same way, and in 0.3.2 they already accept the newline
too:

```
[t][a
b]

[a b]: url
```

```html
<p><a href="url">t</a></p>
```

So the only asymmetry in the released version is the footnote reference rejecting a
newline, which is what your main branch has already removed. Everything else was
uniform all along.

Which suggests the sentence is descriptive rather than a decision:

> A reference label is normalized before lookup, on both the reference and the
> definition side: leading and trailing whitespace is removed and every internal
> run of whitespace, including a line break, becomes a single space. This is why a
> reference that a text editor has wrapped still resolves to a definition written
> on one line.

Stating it also settles the goal 7 question the issue opened with, in the direction
you want: a wrapping tool may break a reference anywhere, and the reference still
binds, while the definition marker stays a one-line construct that a block parser
can find without scanning ahead.

I will move djot-php to match: definitions stay one line, references keep the line
break and normalize. Happy to send the corresponding change here too if that is
useful, though `pattFootnoteStart` is a small enough patch that you may prefer to
do it yourself.
